cxplat/src/cxplat_winuser/rundown_winuser.cpp

```diff
@@ -0,0 +1,209 @@
+// Copyright (c) Microsoft Corporation
+// SPDX-License-Identifier: MIT
+#include "cxplat.h"
+#include <map>
+#include <memory>
+#include <mutex>
+#include <windows.h>
+
+/***
+ * @brief This following class implements a mock of the Windows Kernel's rundown reference implementation.
+ * 1) It uses a map to track the number of references to a given cxplat_rundown_reference_t structure.
+ * 2) The address of the cxplat_rundown_reference_t structure is used as the key to the map.
+ * 3) A single condition variable is used to wait for the ref count of any cxplat_rundown_reference_t structure to reach 0.
+ * 4) The class is a singleton and is created during static initialization and destroyed during static destruction.
```

Comment:
Whis is this so complicated, and not simply a ref count and a completion event? Is it purely for debugging reasons? If so, we need a hyper optimized release mode version.

Reply:
It was already this way in the usersim repo, all I did was rename it and move it to another file.

Reply:
The closest C++ equivalent of a Windows event is a condition variable. The C++ version was used to permit this code to be more easily made cross-platform. 

An event / condition_variable per run-down reference would leak as there is no "free" API in kernel mode.

That said, it we don't care about Linux support, we could switch this to WaitOnAddress/WakeByAddress which is similar to this code, but implemented in the OS.

Reply:
we should move any workarounds for "no free API in kernel mode" over to the usersim directory and keep the complexity out of the cxplat directory.

Reply:
Filed issue #98